üèóÔ∏è Containerized Tools

    DeepVariant

        Poplin, R., et al. (2018). A universal SNP and small indel variant caller using deep neural networks. Nature Biotechnology, 36(10), 983‚Äì987. https://doi.org/10.1038/nbt.4235

    GLnexus

        Lin, M. F., et al. (2018). GLnexus: joint variant calling for large cohort sequencing. bioRxiv, 343970. https://doi.org/10.1101/343970

üêç Conda-Managed Tools

    AdapterRemoval (v2)

        Schubert, M., et al. (2016). AdapterRemoval v2: rapid adapter trimming, identification, and read merging. BMC Research Notes, 9(1), 88. https://doi.org/10.1186/s13104-016-1900-2

    Annovar

        Wang, K., Li, M., & Hakonarson, H. (2010). ANNOVAR: functional annotation of genetic variants from high-throughput sequencing data. Nucleic Acids Research, 38(16), e164. https://doi.org/10.1093/nar/gkq603

    BCFtools / Samtools / HTSlib

        Danecek, P., et al. (2021). Twelve years of SAMtools and BCFtools. GigaScience, 10(2), giab008. https://doi.org/10.1093/gigascience/giab008

    Bedtools

        Quinlan, A. R., & Hall, I. M. (2010). BEDTools: a flexible suite of utilities for comparing genomic features. Bioinformatics, 26(6), 841‚Äì842. https://doi.org/10.1093/bioinformatics/btq033

    Bracken

        Lu, J., et al. (2017). Bracken: estimating species abundance in metagenomics data. PeerJ Computer Science, 3, e104. https://doi.org/10.7717/peerj-cs.104

    Dragmap

        Illumina. (2021). DRAGEN-Map (DRAGMAP). GitHub repository: https://github.com/Illumina/dragmap

    GATK4

        Van der Auwera, G. A., & O'Connor, B. D. (2020). Genomics in the Cloud: Using Docker, GATK, and WDL in Terra. O'Reilly Media. (Note: This is the updated citation for GATK4).

    KMC (v3)

        Kokot, M., Dlugosz, M., & Deorowicz, S. (2017). KMC 3: many-fold faster k-mer counting. Bioinformatics, 33(17), 2759‚Äì2761. https://doi.org/10.1093/bioinformatics/btx304

    Kraken2

        Wood, D. E., Lu, J., & Langmead, B. (2019). Improved metagenomic analysis with Kraken 2. Genome Biology, 20(1), 257. https://doi.org/10.1186/s13059-019-1891-0

    Mosdepth

        Pedersen, B. S., & Quinlan, A. R. (2018). Mosdepth: quick coverage calculation for genomes and exomes. Bioinformatics, 34(5), 867‚Äì868. https://doi.org/10.1093/bioinformatics/btx699

    VerifyBamID2

        Zhang, F., et al. (2020). Ancestry-agnostic estimation of DNA sample contamination from sequence reads. Genome Research, 30(2), 185‚Äì194. https://doi.org/10.1101/gr.246934.118

    WhatsHap

        Martin, M., et al. (2016). WhatsHap: fast and accurate read-based phasing. bioRxiv, 085050. https://doi.org/10.1101/085050

üíª Python Libraries

    Biopython

        Cock, P. J., et al. (2009). Biopython: freely available Python tools for computational molecular biology and bioinformatics. Bioinformatics, 25(11), 1422‚Äì1423.

    Cyvcf2

        Pedersen, B. S., & Quinlan, A. R. (2017). cyvcf2: fast, flexible variant analysis with Python. Bioinformatics, 33(12), 1867‚Äì1869.

    Matplotlib

        Hunter, J. D. (2007). Matplotlib: A 2D graphics environment. Computing in Science & Engineering, 9(3), 90‚Äì95.

    NumPy

        Harris, C. R., et al. (2020). Array programming with NumPy. Nature, 585(7825), 357‚Äì362.

    Pandas

        The pandas development team. (2020). pandas-dev/pandas: Pandas. Zenodo. https://doi.org/10.5281/zenodo.3509134

    Pybedtools

        Dale, R. K., Pedersen, B. S., & Quinlan, A. R. (2011). Pybedtools: a flexible Python library for manipulating genomic datasets and annotations. Bioinformatics, 27(24), 3423‚Äì3424.

    Scikit-learn

        Pedregosa, F., et al. (2011). Scikit-learn: Machine Learning in Python. Journal of Machine Learning Research, 12, 2825‚Äì2830.

    SciPy

        Virtanen, P., et al. (2020). SciPy 1.0: fundamental algorithms for scientific computing in Python. Nature Methods, 17(3), 261‚Äì272.